applications/asset_tracker_v2/src/modules/location_module.c

```diff
@@ -0,0 +1,682 @@
+	int agps_wait_threshold_ms = CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC *
+				     MSEC_PER_SEC;
+
+	if (!IS_ENABLED(CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS)) {
+		return true;
+	} else if (agps_wait_threshold_ms < 0) {
+		/* If CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC is set to -1,
+		 * we need to notify the data module that the application module is awaiting
+		 * A-GPS data in order to request GNSS. If not, GNSS will never be
+		 * requested if the initial A-GPS data request fails.
+		 */
+		if (is_agps_processed()) {
+			return true;
+		}
+
+		SEND_EVENT(app, APP_EVT_AGPS_NEEDED);
+		return false;
+
+	} else if ((agps_wait_threshold_ms < uptime_current_ms) || is_agps_processed()) {
+		return true;
+	}
+
+	return false;
+}
+
+static void search_start(void)
+{
+	int err;
+	struct location_config config;
+	enum location_method methods[2];
+	int methods_count = 0;
+	int methods_index_gnss = -1;
+	int methods_index_cellular = -1;
+
+	if (copy_cfg.no_data.neighbor_cell && copy_cfg.no_data.gnss) {
+		return;
+	}
+
+	if (request_gnss() && !copy_cfg.no_data.gnss) {
+		methods[methods_count] = LOCATION_METHOD_GNSS;
+		methods_index_gnss = methods_count;
+		methods_count++;
+	}
+
+	/* If GNSS shouldn't be requested yet, just request cellular location */
+	if (!copy_cfg.no_data.neighbor_cell) {
+		methods[methods_count] = LOCATION_METHOD_CELLULAR;
+		methods_index_cellular = methods_count;
+		methods_count++;
+	}
+
+	if (methods_count == 0) {
+		LOG_WRN("No location methods selected");
+		return;
+	}
+
+	location_config_defaults_set(&config, methods_count, methods);
+
+	/* The reason for having at least 75 seconds sample timeout when
+	 * requesting GNSS data is that the GNSS module on the nRF9160 modem will always
+	 * search for at least 60 seconds for the first position fix after a reboot. This limit
+	 * is enforced in order to give the modem time to perform scheduled downloads of
+	 * A-GPS data from the GNSS satellites.
+	 *
+	 * However, if A-GPS data has been downloaded via the cloud connection and processed
+	 * before the initial GNSS search, the actual GNSS timeout set by the application is used.
+	 *
+	 * Processing A-GPS before requesting GNSS data is enabled by default,
+	 * and the time that the application will wait for A-GPS data before including GNSS
+	 * in sample requests can be adjusted via the
+	 * CONFIG_APP_REQUEST_GNSS_WAIT_FOR_AGPS_THRESHOLD_SEC Kconfig option. If this option is
+	 * set to -1, the application module will not request GNSS unless A-GPS data has been
+	 * processed.
+	 *
+	 * If A-GPS data has been processed the sample timeout can be ignored as the
+	 * GNSS will most likely time out before the sample timeout expires.
+	 *
+	 * When GNSS is requested, set the sample timeout to (GNSS timeout + 15 seconds)
+	 * to let the GNSS module finish ongoing searches before data is sent to cloud.
+	 */
+	if (methods_index_gnss != -1) {
+		config.methods[methods_index_gnss].gnss.timeout =
```

Comment:
@simensrostad wrote:
`The GNSS timeout received in DATA_EVT_CONFIG_READY and DATA_EVT_CONFIG_INIT events is replaced with a location timeout value. This value is the gnss_timeout present in the cloud_data_cfg structure carried in these events.`

Just want to confirm that do you suggest that we should change gnss_timeout to location_timeout or what do you mean?

Reply:
Yes, gnss_timeout -> location_timeout.

Reply:
Ok. So perhaps do this translation in the location module then? Use a single timeout for both. Or add a second timeout value for cellular timeout. This approach will require that we update support of setting this via the web UI. So updating the codecs for all the cloud implementations and the data module to handle the new config.

Reply:
I hadn't planned to add cellular timeout configuration to ATv2 because it hasn't been configurable so far. We can of course change this if needed.
Currently the implementation is same as earlier for GNSS and ncell measurement timeouts but I'm changing GNSS timeout based on other conversation from:
`MAX(copy_cfg.gnss_timeout + 15, 75)`
To:
`copy_cfg.gnss_timeout`

Cellular timeout has been defined as follows and I didn't plan to change it:
`#define DATA_FETCH_TIMEOUT_NEIGHBORHOOD_SEARCH 11`

Reply:
Ok, how do you notify the location module (API) of the ncell timeout? Maybe just make it kconfigurable in the location module and remove the handling of ncell timeout in main.c?

Reply:
I'm not 100% sure if location module in your text is ATv2 location module or Location library.
Anyway, I'm planning to move `#define DATA_FETCH_TIMEOUT_NEIGHBORHOOD_SEARCH 11` from main.c to ATv2 location module (it's already copied but not removed from main.c yet)., which sets the timeout into location library.

The question mark will be how to set `app_module_event->timeout` in main.c. At the moment it's set based on gnss_timeout always but I need to think about this. I don't know how important the exact value of this timeout is. I guess it's more of a safety guard that we don't wait indefinitely.

Reply:
Sounds good.
Yes, its used as a safeguard. The exact value of `app_module_event->timeout` doesn't really matter, as long as its sufficiently large to give all modules enough time to sample data.

Reply:
This has been changed now so I'll close this discussion.